Stacktrace if I did not specify a Random key for subdocument

I get this:

```
Error: subdocument_key_type_restriction
E             [46] throw(error(subdocument_key_type_restriction,_6238))
E             [44] catch('<garbage_collected>','<garbage_collected>','<garbage_collected>') at /usr/lib/swipl/boot/init.pl:533
E             [42] api_document:api_insert_document_(schema,transaction_object{commit_info:commit_info{},descriptor:branch_descriptor{branch_name:"main",repository_descriptor: ...},inference_objects:[],instance_objects:[...],parent:transaction_object{descriptor: ...,inference_objects:[],instance_objects: ...,parent: ...,schema_objects: ...},schema_objects:[...]},<stream>(0x55bde1421600),_6314) at /app/terminusdb/src/core/api/api_document.pl:175
E             [41] '$bags':findall_loop(_6430,'<garbage_collected>',_6434,[]) at /usr/lib/swipl/boot/bags.pl:99
E             [40] setup_call_catcher_cleanup('$bags':'$new_findall_bag','$bags':findall_loop(_6488,...,_6492,[]),_6470,'$bags':'$destroy_findall_bag') at /usr/lib/swipl/boot/init.pl:614
E             [36] '<meta-call>'('<garbage_collected>') <foreign>
E             [35] database:with_transaction_(query_context{all_witnesses:false,authorization:'terminusdb://system/data/User/admin',bindings:[],commit_info:commit_info{author:admin,message:'I am checking in the schema'},default_collection:branch_descriptor{branch_name:"main",repository_descriptor: ...},files:[],filter:type_filter{types: ...},prefixes:_6650{'@base':"terminusdb:///data/",'@schema':"terminusdb:///schema#",'@type':'Context',api:'http://terminusdb.com/schema/api#',owl:'http://www.w3.org/2002/07/owl#',rdf:'http://www.w3.org/1999/02/22-rdf-syntax-ns#',rdfs:'http://www.w3.org/2000/01/rdf-schema#',sys:'http://terminusdb.com/schema/sys#',vio:'http://terminusdb.com/schema/vio#',woql:'http://terminusdb.com/schema/woql#',xdd:'http://terminusdb.com/schema/xdd#',xsd:'http://www.w3.org/2001/XMLSchema#'},selected:[],system:system_descriptor{},transaction_objects:[...],update_guard:_6620,write_graph:branch_graph{branch_name:"main",database_name:"test_docapi",organization_name:"admin",repository_name:"local",type:instance}},api_document:(...;...),_6574) at /app/terminusdb/src/core/transaction/database.pl:220
E             [34] setup_call_catcher_cleanup(database:true,database:with_transaction_(...,...,_6816),_6794,database:true) at /usr/lib/swipl/boot/init.pl:614
E             [30] '<meta-call>'('<garbage_collected>') <foreign>
E             [29] catch(routes:(...,...),error(subdocument_key_type_restriction,context(_6912,_6914)),routes:do_or_die(...,...)) at /usr/lib/swipl/boot/init.pl:532
E             [28] catch_with_backtrace('<garbage_collected>','<garbage_collected>','<garbage_collected>') at /usr/lib/swipl/boot/init.pl:582
E           
E           Note: some frames are missing due to last-call optimization.
E           Re-run your program in debug mode (:- debug.) to get more detail.
```

When I insert this schema:
```json
[{'@type': 'Class', '@id': 'Address', '@documentation': {'@comment': 'This is address', '@properties': {}}, '@subdocument': [], 'street': 'xsd:string', 'postal_code': 'xsd:string', 'country': 'Country'}, {'@type': 'Class', '@id': 'Country', 'name': 'xsd:string', 'perimeter': {'@type': 'List', '@class': 'Coordinate'}}, {'@type': 'Class', '@id': 'Coordinate', 'x': 'xsd:decimal', 'y': 'xsd:decimal'}, {'@type': 'Enum', '@id': 'Role', '@value': ['Admin', 'Read', 'Write']}, {'@type': 'Class', '@id': 'Employee', '@inherits': ['Person'], 'address_of': 'Address', 'contact_number': {'@type': 'Optional', '@class': 'xsd:string'}, 'managed_by': 'Employee', 'member_of': 'Team', 'permisstion': {'@type': 'Set', '@class': 'Role'}, 'name': 'xsd:string', 'age': 'xsd:integer', 'friend_of': {'@type': 'Set', '@class': 'Person'}}, {'@type': 'Enum', '@id': 'Team', '@value': ['Information Technology', 'Marketing']}, {'@type': 'Class', '@id': 'Person', '@documentation': {'@comment': 'This is a person', '@properties': {'name': 'Name of the person.', 'age': 'Age of the person.'}}, 'name': 'xsd:string', 'age': 'xsd:integer', 'friend_of': {'@type': 'Set', '@class': 'Person'}}]
```

I think how I cause it is to not specify the subdocument to have a "Random" key. I think if I do so it will be Random by default and do not have to specify it?

--- a/src/core/api/api_document.pl
+++ b/src/core/api/api_document.pl
@@ -156,6 +156,9 @@ known_document_error(unrecognized_property(_,_,_)).
 known_document_error(casting_error(_,_)).
 known_document_error(submitted_id_does_not_match_generated_id(_,_)).
 known_document_error(submitted_document_id_does_not_have_expected_prefix(_,_)).
+known_document_error(document_key_type_unknown(_)).
+known_document_error(document_key_type_missing(_)).
+known_document_error(subdocument_key_missing).
 
 :- meta_predicate call_catch_document_mutation(+, :).
 call_catch_document_mutation(Document, Goal) :-

--- a/src/core/api/api_error.pl
+++ b/src/core/api/api_error.pl
@@ -1313,6 +1313,35 @@ api_document_error_jsonld(delete_documents, error(not_a_proper_id_for_deletion(I
                               'api:invalid_document_id' : Id},
              'api:message' : Msg
             }.
+api_document_error_jsonld(Type, error(document_key_type_unknown(Key_Type, Document),_),JSON) :-
+    document_error_type(Type, JSON_Type),
+    format(string(Msg), "Document @key type unknown: ~q. It must be ValueHash, Hash, Lexical, or Random.", [Key_Type]),
+    JSON = _{'@type' : JSON_Type,
+             'api:status' : "api:failure",
+             'api:error' : _{ '@type' : 'api:DocumentKeyTypeUnknown',
+                              'api:key_type': Key_Type,
+                              'api:document' : Document },
+             'api:message' : Msg
+            }.
+api_document_error_jsonld(Type, error(document_key_type_missing(Key, Document),_),JSON) :-
+    document_error_type(Type, JSON_Type),
+    format(string(Msg), "Document @key missing @type: ~q", [Key]),
+    JSON = _{'@type' : JSON_Type,
+             'api:status' : "api:failure",
+             'api:error' : _{ '@type' : 'api:DocumentKeyTypeMissing',
+                              'api:key': Key,
+                              'api:document' : Document },
+             'api:message' : Msg
+            }.
+api_document_error_jsonld(Type, error(subdocument_key_missing(Document),_),JSON) :-
+    document_error_type(Type, JSON_Type),
+    format(string(Msg), "Subdocument @key missing", []),
+    JSON = _{'@type' : JSON_Type,
+             'api:status' : "api:failure",
+             'api:error' : _{ '@type' : 'api:SubdocumentKeyMissing',
+                              'api:document' : Document },
+             'api:message' : Msg
+            }.
 
 /**
  * generic_exception_jsonld(Error,JSON) is det.

--- a/src/core/document/json.pl
+++ b/src/core/document/json.pl
@@ -893,6 +893,13 @@ json_schema_elaborate_key(V,_,json{ '@type' : Type}) :-
     get_dict('@type', V, Random),
     expand_match_system(Random, 'Random', Type),
     !.
+json_schema_elaborate_key(V,_,_) :-
+    get_dict('@type', V, Type),
+    !,
+    throw(error(document_key_type_unknown(Type), _)).
+json_schema_elaborate_key(V,_,_) :-
+    atom_json_dict(Atom, V, []),
+    throw(error(document_key_type_missing(Atom), _)).
 
 json_schema_elaborate_property_documentation(Context, Path, Dict, Out) :-
     global_prefix_expand(sys:'PropertyDocumentation',Property_Ex),
@@ -1074,18 +1081,21 @@ check_schema_document_restrictions(Elaborated) :-
     get_dict(AbsP, Elaborated, _),
     !.
 check_schema_document_restrictions(Elaborated) :-
+    global_prefix_expand(sys:key, KeyP),
+    do_or_die(get_dict(KeyP, Elaborated, Key),
+              error(subdocument_key_missing, _)),
+    % We currently don't reach the following checks because
+    % json_schema_elaborate_key throw errors for the same conditions.
+    do_or_die(get_dict('@type', Key, Key_Type_String),
+              error(subdocument_key_type_missing,_)),
+    atom_string(Key_Type, Key_Type_String),
     do_or_die(
-        (   global_prefix_expand(sys:key, KeyP),
-            get_dict(KeyP, Elaborated, Key),
-            % Is this not exhaustive?
-            (   global_prefix_expand(sys:'ValueHash',Key_Type)
-            ;   global_prefix_expand(sys:'Hash',Key_Type)
-            ;   global_prefix_expand(sys:'Lexical',Key_Type)
-            ;   global_prefix_expand(sys:'Random',Key_Type)),
-            get_dict('@type', Key, Key_Type_String),
-            atom_string(Key_Type, Key_Type_String)
-        ),
-        error(subdocument_key_type_restriction,_)).
+        % Is this exhaustive?
+        (   global_prefix_expand(sys:'ValueHash',Key_Type)
+        ;   global_prefix_expand(sys:'Hash',Key_Type)
+        ;   global_prefix_expand(sys:'Lexical',Key_Type)
+        ;   global_prefix_expand(sys:'Random',Key_Type)),
+        error(subdocument_key_type_unknown(Key_Type_String),_)).
 
 json_schema_elaborate(JSON,Context,JSON_Schema) :-
     json_schema_elaborate(JSON,Context,[],JSON_Schema).

